### PR TITLE
AEROGEAR-2091 - JWT validation implementation

### PR DESCRIPTION
**Pull request by rachael-oregan**
### Motivation

JIRA: https://issues.jboss.org/browse/AEROGEAR-2091

### Description
The [JSONWebToken](https://github.com/kreactive/JSONWebToken) library is used to verify JSON Web Tokens.  I'm using `JSONWebToken`@v2.0.0 because the latest version of the library (2.1.0) requires Swift v4.1.0 and our project uses 4.0.0.  

### Progress

- [x] `verifyClaims` function
- [x] unit tests
- [x] docs

### CODE REVIEW

**Comment by wtrocki** [comment]
@rachael-oregan Community tend to update version rapidly. We should do that as well since we do not have official release. Feel free to bump the version as it should not cause any issues. If somehow you do not want to do it in this PR let me know. I will do that in separate one and we could rebase your PR. Have fun.

File: modules/auth/jwt/Jwt.swift [hunk 1]
```diff
@@ -22,19 +16,21 @@ func base64Decode(_ input: String) -> Data? {
 
     if base64.count % 4 != 0 {
         base64.append(String(repeating: "=", count: 4 - base64.count % 4))
+    }
+    if (Data(base64Encoded: base64) == nil) {
```
Comment (wei-lee) [inline modules/auth/jwt/Jwt.swift]: is this `if` statement required?

File: modules/auth/jwt/Jwt.swift [hunk 2]
```diff
@@ -48,20 +44,51 @@ class Jwt {
...
+            
+            let validationResult = validator.validateToken(jwt)
+            
+            if (validationResult.isValid) {
+                return true
+            } else {
+                return false
+            }
+        } catch {
+            throw Errors.generatingPublicKey("Error creating RSA public Key using the provided modulus and exponent", error)
```
Comment (wei-lee) [inline modules/auth/jwt/Jwt.swift]: the error message is specific about generating public key. It should either catch the specific error threw by `RSAKey.registerOrUpdateKey`,  or the catch block should be moved up after line 74.

File: modules/auth/jwt/Jwt.swift [hunk 3]
```diff
@@ -48,20 +44,51 @@ class Jwt {
...
+     
+     - parameters:
+        - jwks: the JSON Web Key Set to verify the JWT against
+        - jwt: the JSON Web Token to be verified
+     
+     - throws: an `invalidModulus` error if the decoded modulus is nil, an `invalidExponent` error if the decoded exponent is nil or a `generatingPublicKey` error if the RSA public key could not be generated
+     
+     - returns: true if the JWT is valid, false otherwise
+     */
+    public static func verifyJwt(jwks: Jwks, jwt: String) throws -> Bool {
```
Comment (wei-lee) [inline modules/auth/jwt/Jwt.swift]: so where this function is being used? I think the `AuthService` module should be updated to actually use this method to verify the token?

File: modules/auth/jwt/Jwt.swift [hunk 4]
```diff
@@ -48,20 +44,51 @@ class Jwt {
...
+        guard let modulus = base64Decode(jwks.keys[0].n) else {
```
Comment (wei-lee) [inline modules/auth/jwt/Jwt.swift]: so the keys in `jwks` is an array. However, depending on how many keys that are available, there might be multiple keys in the array and it's not guaranteed the first one will be a RSA type key. It should be a AES type of key.

So we should do a lookup in the array and use the first key whose type is `RSA`.

File: modules/auth/jwt/Jwt.swift (hunk 1 above)
Comment (rachael-oregan) [reply modules/auth/jwt/Jwt.swift]: Ah I must have forget to remove this, thanks

File: modules/auth/jwt/Jwt.swift (hunk 2 above)
Comment (rachael-oregan) [reply modules/auth/jwt/Jwt.swift]: Ya that makes sense, I'll update this to catch the `badKeyFormat` error threw by the `registerOrUpdateKey` function

File: modules/auth/jwt/Jwt.swift (hunk 3 above)
Comment (rachael-oregan) [reply modules/auth/jwt/Jwt.swift]: I'll update the `AuthService` to verify the tokens

File: modules/auth/jwt/Jwt.swift (hunk 4 above)
Comment (rachael-oregan) [reply modules/auth/jwt/Jwt.swift]: Ah ok,  I thought there is usually only one key returned here.  I'll update this to find the `RSA` key.  Thanks for pointing that out

**Comment by rachael-oregan** [comment]
@wtrocki thanks for having a look. Ya that makes sense to bump the version.  However, to get latest version of Swift (4.1.0) I'll need to update to macOS High Sierra 10.13.4.  I can update this this evening after work cause it'll take a while.

If you have the latest version then feel free to bump the version too.  Thanks :)

**Comment by rachael-oregan** [comment]
@wei-lee mind having another look please. Thanks

File: modules/auth/jwk/JwksManager.swift [hunk 5]
```diff
@@ -2,6 +2,41 @@ import AGSCore
...
+ - returns: the RSA JSON Web Key or nil if not found.
+ */
+public func rsaJwk(jwks: Jwks) -> JwkContent? {
+    var rsaJwk: JwkContent?
+    for (_, jwk) in jwks.keys.enumerated() {
+        if (jwk.kty == "RSA") {
+            rsaJwk = jwk
+        }
+    }
+    return rsaJwk
```
Comment (wei-lee) [inline modules/auth/jwk/JwksManager.swift]: I just read about how to lookup keys again, looks like we need a small improvement here:

apparently in the JWT token, there is a `kid` header, which gives you the id of the key that is used to sign/encrypt the token.

That value should be used to look up the key. So a small change to this function is required. It should take another parameter, which is the `kid` value in the JWT token, then the lookup logic should be to changed to `jwt.kid == kid`.

File: modules/auth/jwt/Jwt.swift [hunk 6]
```diff
@@ -48,20 +42,54 @@ class Jwt {
...
+        }
+        guard let exponent = base64Decode(jwk.e) else {
+            throw Errors.invalidExponent("Decoded exponent cannot be nil")
         }
 
-        return JSONWebToken(header: headerValue, payload: payload!, signature: signatureValue)
+        do {
+            let publicKey: RSAKey = try RSAKey.registerOrUpdateKey(modulus: modulus, exponent: exponent, tag: "publicKey")
+              
+            let jwt = try Jwt.decode(jwt)
```
Comment (wei-lee) [inline modules/auth/jwt/Jwt.swift]: so this line should be moved to, which you can then access the `kid` value in the headers of the token.

Then you can use the `kid` value to lookup the right key in the `JWKS`.

**Comment by rachael-oregan** [comment]
@wtrocki if you have some time would you mind reviewing these changes please?

File: example/AeroGearSdkExampleTests/auth/jwk/JwkManagerTests.swift [hunk 7]
```diff
@@ -23,7 +23,7 @@ class JwkManagerTests: XCTestCase {
 
         // create JWKS
         let jwks = try? JSONSerialization.data(withJSONObject: getMockJwks(), options: [])
-        let jwksString = String(data: jwks!, encoding: String.Encoding.ascii)
+        let jwksString = String(data: jwks!, encoding: .utf8)
```
Comment (wtrocki) [inline example/AeroGearSdkExampleTests/auth/jwk/JwkManagerTests.swift]: I will avoid `jwks!` 
We should unwrap optional.

File: modules/auth/AgsAuth.swift [hunk 8]
```diff
@@ -146,15 +148,29 @@ open class AgsAuth {
...
         guard configured else {
             throw Errors.serviceNotConfigured
         }
+        
         guard let currentCredential = credentialManager.load() else {
             return nil
         }
-        guard let user = User(credential: currentCredential, clientName: keycloakConfig!.clientID) else {
+        
+        guard let jwks = jwksManager?.load(keycloakConfig!) else {
```
Comment (wtrocki) [inline modules/auth/AgsAuth.swift]: keycloakConfig! can be unwrapped above and used directly as non optional. `!` is bad as if under some circumstances it's nil app will crash

File: modules/auth/jwk/JwksManager.swift [hunk 9]
```diff
@@ -80,12 +117,17 @@ class JwksManager {
                     return onCompleted!(nil, error)
                 }
             } else if let response = response.response as? [String: Any] {
+                var responseData: Data?
                 if let resData = try? JSONSerialization.data(withJSONObject: response, options: []) {
+                    responseData = resData
                     let resString = String(data: resData, encoding: .utf8)
                     self.persistJwks(keycloakConfig.realmName, resString!)
                 }
                 if onCompleted != nil {
```
Comment (wtrocki) [inline modules/auth/jwk/JwksManager.swift]: Unwrap this instead of comparing to nil.

File: modules/auth/jwt/Jwt.swift (hunk 3 above)
Comment (wtrocki) [reply modules/auth/jwt/Jwt.swift]: @rachael-oregan Was that done? Cannot see this

File: modules/auth/AgsAuth.swift (hunk 8 above)
Comment (rachael-oregan) [reply modules/auth/AgsAuth.swift]: The `configured` var is used to do this check [here](https://github.com/rachael-oregan/aerogear-ios-sdk/blob/ff07f0e83d05a8c12702271083663f57aaf3f136/modules/auth/AgsAuth.swift#L46)

File: modules/auth/jwt/Jwt.swift (hunk 3 above)
Comment (rachael-oregan) [reply modules/auth/jwt/Jwt.swift]: Yes it's been used to verify the token [here](https://github.com/rachael-oregan/aerogear-ios-sdk/blob/ff07f0e83d05a8c12702271083663f57aaf3f136/modules/auth/AgsAuth.swift#L169)

**Comment by rachael-oregan** [comment]
@wtrocki I refactored to unwrap the callback optionals.  Kinda looks a bit ugly with all the `if let`s.  Let me know if this can be improved

**Pull request merged**